### PR TITLE
Add submission files.

### PR DESCRIPTION
**Pull request by athon-millane**

### CODE REVIEW

File: athon_millane/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Skill Set Challenge
+---
```
Comment (PanPip) [inline athon_millane/README.md:2]: A very nice readme and descriptions 👍

File: athon_millane/data/utils.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+def pull_sp500(force=False):
+    """
+    Pull S&P500 daily close prices for the past 20 years.
+
+    :param force (bool): If true, overwrite existing data file.
+    """
+
+    # Check if data already exists.
+    if path.exists(DATA_PATH) and not force:
+        return
```
Comment (PanPip) [inline athon_millane/data/utils.py:23]: Docstrings and inline comments match our code style. Perfect.

File: athon_millane/parameter_sweep.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        :param returns_df (pd.DataFrame): Returns df to run experiments on.
+        """
+
+        if returns_df is not None:
+            self.returns_df = returns_df
+
+        else:
+            if not simulate:
+                raise Exception(
+                    "Please provide a returns df or set simulate=True")
```
Comment (PanPip) [inline athon_millane/parameter_sweep.py:48]: I guess this part was not implemented or the `simulate` parameter is not added.

File: athon_millane/parameter_sweep.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+                           values=args[2])
+            ann = pd.DataFrame(d.values >= d.values.max()
+                               ).replace({True: "X", False: ""})
+            sns.heatmap(d, annot=ann, fmt="s", **kwargs)
+
+        # Create a facetgrid and map the heatmap to that grid, with nice colour
+        # mapping.
+        grid = sns.FacetGrid(experiment_df, row='j', col='kappa')
+        cbar_ax = grid.fig.add_axes([1.03, .15, .03, .7])
+        cmap = sns.color_palette("Spectral", as_cmap=True)
```
Comment (PanPip) [inline athon_millane/parameter_sweep.py:74]: Great job on improving plot visuals and style.

File: athon_millane/parameter_sweep.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+            plot: bool = True) -> pd.DataFrame:
+        """
+        Sweep across all parameters of interest for a policy on a single
+        (real) portfolio. Experiment results are visualised as a grid
+        of heatmaps covering all parameters of the stop loss process.
+
+        :param sweep_params: (dict): Dictionary of sweep parameters.
+        :param metric: (str): Performance measure for sweep.
+        :param plot: (bool): Flag for plotting results.
+        :return (pd.DataFrame): Returns results table if plot=False.
```
Comment (PanPip) [inline athon_millane/parameter_sweep.py:98]: This method would return `pd.DataFrame` in any case, just a minor detail.

File: athon_millane/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+yfinance
+yahoo_fin
```
Comment (PanPip) [inline athon_millane/requirements.txt:2]: Good that you've added a list of requirements. The versions could be specified.

File: athon_millane/parameter_sweep.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    def __init__(self, s: int, rho: int):
+        """
+        Constructor for process.
+
+        :param s: (int): Scale factor.
+        :param rho: (int): Autoregressive parameter.
+        """
+        self.ar = np.array([s, -s * rho])
+        self.ma = np.array([1])
+        self.process = ArmaProcess(self.ar, self.ma)
```
Comment (PanPip) [inline athon_millane/parameter_sweep.py:296]: This works 🙂

File: athon_millane/stop_loss_policy.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        This method applies the stop loss policy to the portfolio
+        and generates a state sequencyaccording to definition 1 of the paper.
+        If kappa is provided, transaction cost will be considered in all
+        policy return calculations. If rolling_sd is set to True, gamma and
+        delta are considered as multiples of a rolling standard deviation,
+        otherwise their value is held constant.
+
+        :param gamma: (float): The stopping threshold.
+        :param delta: (float): The re-entry threshold.
+        :param j: (float): The rolling window size for R(J) and rolling sigma.
```
Comment (PanPip) [inline athon_millane/stop_loss_policy.py:138]: Why not ask for an `int` parameter?

File: athon_millane/stop_loss_policy.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+            elif row["r_t"] < row["delta"] and state == 0:
+                state = 0
+
+            states.append(state)
+        return states[:-1]
+
+    def _error_check(self, check_applied=False):
+        """
+        Helper method to run error checks and ensure methods are called
+        in the correct order.
```
Comment (PanPip) [inline athon_millane/stop_loss_policy.py:107]: It's nice that you've added this method.

File: athon_millane/stop_loss_policy.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+
+        measures = []
+
+        measures.append(("Stopping Premium", stopping_premium))
+        measures.append(("Stopping Ratio", stopping_ratio))
+        measures.append(("Variance Difference", variance_difference))
+        measures.append(("Standard Deviation Difference", sd_difference))
+        measures.append(("Sharpe Difference", sharpe_difference))
+
+        return pd.DataFrame(measures)
```
Comment (PanPip) [inline athon_millane/stop_loss_policy.py:262]: A great output format 👍

File: athon_millane/test.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import unittest
+
+import pycodestyle as pep8
+
+
+class TestCodeFormat(unittest.TestCase):
```
Comment (PanPip) [inline athon_millane/test.py:6]: It's good to see a style unit test. More can be added to test the functionality, error handling, and some edge cases.

File: athon_millane/vectorised_policy_utils.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import numpy as np
+from numba import njit
+
+
+@njit
+def policy_step_jit(ret: float, rolling_j: float, state_: int,
+                    gamma: float, delta: float) -> int:
```
Comment (PanPip) [inline athon_millane/vectorised_policy_utils.py:7]: A very nice extra step to speed up the code!

File: athon_millane/Submission.ipynb [hunk 13]
```diff
@@ -0,0 +1,755 @@
...
+    "\n",
+    "# Implementation of a Stop Loss Policy with Empirical Results\n",
+    "\n",
+    "## Introduction\n",
+    "This notebook demonstrates an implementation of the Stop Loss Policy from the paper, *When Do Stop-Loss Rules Stop Losses?*, Kaminski, Lo.\n",
+    "\n",
+    "In the paper, the authors define a simple Stop-Loss Policy $\\mathcal{S}(\\gamma, \\delta, J, \\kappa)$ for a portfolio strategy $P$ with returns $\\{r_t\\}$. This is a dynamic binary asset-allocation rule $\\{s_t\\}$ between $P$ and a risk-free asset $F$ with return $r_f$ , where $s_t$ is the proportion of assets allocated to $P$, and:\n",
+    "\n",
+    "$s_{t} \\equiv\\left\\{\\begin{array}{llllll}0 & \\text { if } & R_{t-1}(J) & <-\\gamma & \\text { and } & s_{t-1}=1 & \\text { (exit) } \\\\ 1 & \\text { if } & r_{t-1} & \\geq \\delta & \\text { and } & s_{t-1}=0 & \\text { (re-enter) } \\\\ 1 & \\text { if } & R_{t-1}(J) & \\geq-\\gamma & \\text { and } & s_{t-1}=1 & \\text { (stay in) } \\\\ 0 & \\text { if } & r_{t-1} & <\\delta & \\text { and } & s_{t-1}=0 & \\text { (stay out) }\\end{array}\\right.$\n",
```
Comment (PanPip) [inline athon_millane/Submission.ipynb:19]: Good use of LaTeX in the notebook.

File: athon_millane/Submission.ipynb [hunk 14]
```diff
@@ -0,0 +1,755 @@
...
+    "To properly ameleriate this issue, a shorter time period would likely be necessary to allow for faster reaction to market conitions. To investigate whether better parameters can be found at the daily timeframe, we implement a parameter sweep."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "86cf89fb-7c98-4dac-8c41-28af143a9ff4",
+   "metadata": {},
+   "source": [
+    "### Vectorised implementation\n",
+    "Our API provides a JIT compiled implementation, shown here to provide a **70x** speed up over serial policy iteration."
```
Comment (PanPip) [inline athon_millane/Submission.ipynb:290]: The speed improvement is impressive.

File: athon_millane/Submission.ipynb [hunk 15]
```diff
@@ -0,0 +1,755 @@
...
...QuAnudYY/PmzcjPz8err76KxYsXo127dvj8888xf/58rF+/nnPWTUpKwrhx4/DTTz+ZXGJrQ8IyoYiNGzfiyJEjAHw2bv4HLDs7G7NmzUJNTQ3Wrl0Ll8uF66+/HmVlZVi7di3279+PH3/8kSJjxAjh+nn27Nmorq7Gp59+CpvNhg8//JC7r7a2FhMmTDClzIRyNm7ciKNHj3L/HzlyBAsXLkRSUhJ++uknnD59mvtty5YtSE1NxaBBg/D6669j7ty5FAUlRgjXz5s3b0ZVVRUGDx6MgQMH4pNPPuGuO3XqFIYOHWpGkQkVbNy4EYcPHwYAlJSUoLKyEgCQmpqKffv2AQA6duyI8ePH48477+TuO378OHr37m18gWMIcvAjZLF3717ceOONyMnJAcMwOPfcczF69GhkZ2fjv//9LyZNmoSUlBRcddVV+Ne//oXLL78cc+fORW5uLgDg0UcfRUZGBm3lWRy5/XzllVfilVdewTPPPIPly5fjzjvvxO7du5Geno65c+eaXQ0iAvx+ttlsmDRpEqZOnYp27dph8ODB+Oqrr/Dhhx+ie/fuOPvsswH4NFX79+/H5Zdfjri4ODzwwAPo3LmzyTUhwiGnnz/44AP06NEDQ4cOxdKlSzFr1iwsWbIEu3fvRmpqKglRMYDUvD18+HDU1NTA5XLh1KlTSE9PBwDcdtttmD9/PpYsWYKdO3ciNzcXBQUF9H4OAwnLRFgaGhrgcrnwww8/YMaMGbj88svxv//9D19//TUOHz6MWbNm4c033+S2Zjt16oT8/Hx07twZubm58Hg8sNlsmDp1qsk1IcKhtJ+7dOmC1NRUNDQ04NVXX8WhQ4dQUVGBESNGmFwTIhxi/fzZZ5/hiy++wPHjx3H11VcDAEaNGoUNGzbghx9+QMeOHVFQUACv14vKykpceeWVGDZsmMk1IcKhtJ83bdqE9u3bo23btnjppZdw4sQJlJWVkVbZ4oSbt48cOYIbbrgBgM+Z0+v1YuTIkdy9L774Ivbu3YvS0lKcddZZJtUgdiBhmRClqakJTz31FEpKSnDuuefiyy+/5LTEZ599NuLi4vCvf/0L/fr1Q1ZWFh577DEsWrQICQkJqKysREFBAQA6kMTqqO3n+Ph41NXVISEhAfHx8cjKyjK5JkQ4IvWz2+3GP/7xD4wePRrdunUDAEyZMgUrVqzApk2bk
```
Comment (PanPip) [inline athon_millane/Submission.ipynb:586]: A good comparison here.

Comment (PanPip) [inline athon_millane/Submission.ipynb:709]: The notebook is also according to our style guidelines 👍

**Review by PanPip — COMMENTED** [review]
Hi Athon,

This is an impressive piece of work. The code and the notebook follow our style guidelines. I like how you went the extra mile with JIT and improved visualisations.

Will set up an interview.

* Good descriptions of the development process.
* Notebook descriptions are detailed.
* Plots have good style and are easy to comprehend.
* The results of the original paper were replicated and tested.
* Code could be a bit more readable if a higher max line length was used.
* Good separation of code into classes.
* Good private methods usage.
* Code style unit tests are added. Having some functionality unit tests would be amazing.
* Used JIT compilation.
* Added error handling.
* Good use of other packages.
